File: webapp/app/js/controllers/streamingConfig.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

KylinApp.controller('streamingConfigCtrl', function ($scope,StreamingService, $q, $routeParams, $location, $window, $modal, MessageService, CubeDescService, CubeService, JobService, UserService, ProjectService, SweetAlert, loadingRequest, $log, modelConfig, ProjectModel, ModelService, MetaModel, modelsManager, cubesManager, TableModel, $animate,StreamingModel) {

  $scope.tableModel = TableModel;

  if($scope.state.mode=='view') {
    $scope.streamingMeta = StreamingModel.createStreamingConfig();
    $scope.kafkaMeta = StreamingModel.createKafkaConfig();
  }

  if($scope.state.mode=='edit'&& $scope.state.target=='kfkConfig' && $scope.state.tableName){
    StreamingService.getConfig({table:$scope.state.tableName, project: ProjectModel.selectedProject}, function (configs) {
      if(!!configs[0]&&configs[0].name.toUpperCase() == $scope.state.tableName.toUpperCase()){
        $scope.updateStreamingMeta(configs[0]);
        StreamingService.getKfkConfig({kafkaConfigName:$scope.streamingMeta.name, project: ProjectModel.selectedProject}, function (streamings) {
          if(!!streamings[0]&&streamings[0].name.toUpperCase() == $scope.state.tableName.toUpperCase()){
            $scope.updateKafkaMeta(streamings[0]);
          }
        })
      }
    })
  }


  $scope.addCluster = function () {
    $scope.kafkaMeta.clusters.push(StreamingModel.createKafkaCluster());
  };

  $scope.removeCluster = function(cluster){

    SweetAlert.swal({
      title: '',
      text: 'Are you sure to remove this cluster ?',
      type: '',
      showCancelButton: true,
      confirmButtonColor: '#DD6B55',
      confirmButtonText: "Yes",
      closeOnConfirm: true
    }, function(isConfirm) {
      if(isConfirm) {
        var index = $scope.kafkaMeta.clusters.indexOf(cluster);
        if (index > -1) {
          $scope.kafkaMeta.clusters.splice(index, 1);
        }
      }

    })
  }

  $scope.addBroker = function (cluster,broker) {
    //$scope.modelsManager.selectedModel = model;
	if(!!broker){
	  cluster.newBroker = {};
	  angular.copy(broker, cluster.newBroker);
	  cluster.edit_index = cluster.brokers.indexOf(broker);
	}else{
	  cluster.newBroker = StreamingModel.createBrokerConfig();
	}
  };

  $scope.removeNewBroker = function (cluster){
    $scope.clearNewBroker(cluster);
  }

  $scope.removeElement = function (cluster, element) {
    var index = cluster.brokers.indexOf(element);
    if (index > -1) {
      cluster.brokers.splice(index, 1);
    }
    $scope.clearNewBroker(cluster);
  };

  $scope.saveNewBroker = function(cluster){
    if ($scope.isIDExistentInArray(cluster.brokers,cluster.newBroker)) {
      if (cluster.edit_index > -1) {
        cluster.brokers.splice(cluster.edit_index, 1, cluster.newBroker);
      }else{
        SweetAlert.swal('', 'The new record ID already exists.', 'warning');
        return;
      }
	}else{
      cluster.brokers.push(cluster.newBroker);
    }
    $scope.clearNewBroker(cluster);
  }

  $scope.isIDExistentInArray = function(arr,obj){
      if(!arr || arr.length === 0){
        return false;
      }
      for(var i=0; i<arr.length; i++) {
        if(arr[i].id == obj.id){
            return true;
        }
      }
      return false;
    }

  $scope.clearNewBroker = function(cluster){
	delete cluster.edit_index;
    delete cluster.newBroker;
  }


  $scope.streamingTsColUpdate = function(){
    if(!$scope.streamingCfg.parseTsColumn){
      $scope.streamingCfg.parseTsColumn = ' ';
    }
    $scope.kafkaMeta.parserProperties = "tsColName="+$scope.streamingCfg.parseTsColumn;
  }

  $scope.$watch('tableModel.selectedSrcTable', function (newValue, oldValue) {
    if (!newValue) {
      return;
    }
    //view model
    if($scope.state.mode == 'view' && $scope.tableModel.selectedSrcTable.source_type==1){
      $scope.reloadMeta();
    }

  });

  $scope.$on('StreamingConfigEdited', function (event) {
    $scope.reloadMeta();
  });

  $scope.reloadMeta = function(){
    var table = $scope.tableModel.selectedSrcTable;
    var streamingName = table.database+"."+table.name;
    $scope.streamingMeta = {};
    $scope.kafkaMeta = {};
    StreamingService.getConfig({table:streamingName, project: ProjectModel.selectedProject}, function (configs) {
      if(!!configs[0]&&configs[0].name.toUpperCase() == streamingName.toUpperCase()){
        $scope.streamingMeta = configs[0];
        StreamingService.getKfkConfig({kafkaConfigName:$scope.streamingMeta.name, project: ProjectModel.selectedProject}, function (streamings) {
          if(!!streamings[0]&&streamings[0].name.toUpperCase() == streamingName.toUpperCase()){
            $scope.kafkaMeta = streamings[0];
          }
        })
      }
    })
  }

});

File: src/core-metadata/src/main/java/org/apache/kylin/metadata/streaming/StreamingConfig.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kylin.metadata.streaming;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import org.apache.kylin.common.persistence.JsonSerializer;
import org.apache.kylin.common.persistence.ResourceStore;
import org.apache.kylin.common.persistence.RootPersistentEntity;
import org.apache.kylin.common.persistence.Serializer;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 */
@SuppressWarnings("serial")
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.NONE, getterVisibility = JsonAutoDetect.Visibility.NONE, isGetterVisibility = JsonAutoDetect.Visibility.NONE, setterVisibility = JsonAutoDetect.Visibility.NONE)
public class StreamingConfig extends RootPersistentEntity {

    public static Serializer<StreamingConfig> SERIALIZER = new JsonSerializer<StreamingConfig>(StreamingConfig.class);

    public static final String STREAMING_TYPE_KAFKA = "kafka";

    @JsonProperty("name")
    private String name;

    @JsonProperty("type")
    private String type = STREAMING_TYPE_KAFKA;

    @Override
    public String resourceName() {
        return name;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    @Override
    public String getResourcePath() {
        return concatResourcePath(name);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public static String concatResourcePath(String name) {
        return ResourceStore.STREAMING_RESOURCE_ROOT + "/" + name + ".json";
    }

    @Override
    public StreamingConfig clone() {
        try {
            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
            SERIALIZER.serialize(this, new DataOutputStream(baos));
            return SERIALIZER.deserialize(new DataInputStream(new ByteArrayInputStream(baos.toByteArray())));
        } catch (IOException e) {
            throw new RuntimeException(e);//in mem, should not happen
        }
    }

}
